src/shogun/lib/SGMatrix.h

```diff
@@ -306,9 +308,66 @@ template<class T> class SGMatrix : public SGReferencedData
 		/** Set matrix to a constant */
 		void set_const(T const_elem);
 
-		/** fill matrix with zeros */
+		/** Fill matrix with zeros */
 		void zero();
 
+		/**
+		 * Fill matrix with uniformly distributed randoms
+		 * Float Types: [0, 1)
+		 * Int Types: std::numeric_limit<T>::(min->max)
+		**/
+		// vectorized available
+		template <typename U>
+		using enable_simd_float = std::enable_if_t<is_any_of_v<U, float64_t>>;
+		template <typename U = T>
+		auto random() -> enable_simd_float<U>
+		{
+			CRandom r{};
```

Comment:
one thing: we usually allocate sgobjects using `new` for several reasons. It is a style thing but let's adopt that here as well

Reply:
Umm, sure, I was trying following the [guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-scoped).
Also I guess instead of `new` I would need to `std::make_unique` since we want to avoid `SG_UNREF`?

Reply:
Nice questions, @saatvikshah1994 :-)

I'd say the actual reasons are mostly due to history. Shogun has been around for long time, longer than what we consider today modern C++. Also, I personally think the focus and interests have also moved a bit from "hardcore" machine learning programming to software engineering.

It is of course always good to follow the latest trends and guidelines of C++. I've found however that as a project grows and gets bigger such redesigns and/or refactorings become too time consuming without having a clear value in terms of bringing new features. Technical debt.

Reply:
About your comment on `SGMatrix`'s API. Note that the main SG* guys, `SGMatrix` and `SGVector` are referenced data, so there's no need for handmade memory management with those ;-)

Reply:
About the initialization of `CRandom` that Heiko mentioned. It has to do with objects residing in the stack vs the heap. The rule-of-thumb I started using at some moment was that any `SGObject` has to be heap-allocated.

I forgot the exact technical reason. Stepping through a minimal program with gdb could help understanding. But anyway, very vaguely from what I can remember: it had to do with the reference counting mechanism taking place as scopes end.

Anyhow, I think should be able to follow the CPP core guidelines for this initialization: just wrap your `SGObject` inside a `shogun::Some` (aka `std::shared_ptr`).